app/shared/json_convertible.rb

```diff
@@ -224,6 +224,24 @@ def _from_json!(var, val, is_iterable: false)
 
         return var_name, var_value
       end
+
+      def _initialize_using!(json_object)
+        instance = allocate
+        required_init_params = instance.method(:initialize).parameters
+                                       .select { |arg| arg[0] == :keyreq }
+                                       .map(&:last)
+        unless (required_init_params - json_object.keys).empty?
+          raise "Required initialization parameters not found in the object: #{json_object}"
+        end
+        init_params_hash = json_object.select { |key, _value| required_init_params.include?(key) }
+        if instance.method(:initialize).parameters.empty?
+          instance.send(:initialize)
+        else
+          instance.send(:initialize, init_params_hash)
+        end
+        clean_json_object = json_object.reject { |key| required_init_params.include?(key) }
+        return instance, clean_json_object
+      end
```

Comment:
This is some impressive Ruby hacking 🤯 but to be honest I am having a hard time groking it. Is there something we can do that's simpler?

Reply:
Perhaps, enforce that convertible classes must optional parameters?